src/audio/clap/plugin_manager.cpp

```diff
@@ -168,28 +75,31 @@ void PluginManager::EnumeratePlugins()
 			LOG_DEBUG("CLAP: Trying to load plugin library '%s'",
 			          library_path.string().c_str());
 
-			const auto [_, plugin_entry] = load_plugin_library(library_path);
+			std::shared_ptr<Library> library = nullptr;
 
-			if (!plugin_entry) {
-				LOG_WARNING("CLAP: Invalid plugin library '%s'",
-				            library_path.string().c_str());
+			try {
+				library = std::make_shared<Library>(library_path);
+			} catch (const std::runtime_error& ex) {
```

Comment:
> I think this is safe the way everything is currently written. There's a bit of a hidden footgun here though in that `EnumeratePlugins()` must only be called once and also before you load any library. Just something to watch out for if you or someone else refactors this in the future.

`EnumeratePlugins()` is private, so it's called only once. But there's no harm in calling it multiple times. We could listen to "directory changed" async filesystem events and invalidate the cache, but it's not really worth it.

Well, `LoadPlugin()` expects a `PluginInfo` struct, and you can only get the plugin infos from `GetPluginInfos()` that internally calls `EnumeratePlugins()` once (or returns the cached results). I don't see anything particularly fragile here or any hidden dangers (just the "dangers" that you have to face when touching any non-trivial code).

> The Library created here goes out of scope at function end so it does global init + deinit then another global init + deinit when you actually want to use that library. This is fine but I just wanted to point that out.

100% intentional and by design. We need to load the library and intialise it for enumerating the plugins, but we don't want to keep all loaded libraries in memory just because we enumerated them. That's the way to do plugin enumeration.

Real-life audio hosts always cache the enumerated results on the filesystem. So if you have lots of plugins and you use a DAW for the first time, you can wait potentially for minutes until the plugin discovery happens by loading & unloading plugins. This is also necessary to blacklist crash-prone plugins that for some reason don't work in your DAW--better to blacklist them at discovery stage rather than crash in the middle of your unsaved session when you actually want to use them!

Subsequent DAW starts will be fast and the library load/unload doesn't happen; it's working from the cached results stored in a file.  E.g., these are the CLAP and VST plugin preferences in REAPER (my DAW of choice). Note the "Re-scan" buttons and the "Scan new/updated plug-ins at startup" for VSTs (seems like it's automatic for CLAP). So yeah, this is the standard approach.

![image](https://github.com/user-attachments/assets/c86227e1-3cb9-4a3d-81b1-a544636983a7)

![image](https://github.com/user-attachments/assets/3e017b7a-5235-4ba1-8890-73e53a9d3c1e)

> I would have considered keeping owned objects around in the cache but since you've got `PluginManager` as a static object, that's probably a bad idea.

As above, we don't need to keep them in the cache. Discovery loads/unloads, that's fine. Instantiation is different.

> Also, none of this stuff is thread safe but I think you're only running this on the main thread (I hope?)

It's not meant to be thread-safe, most of our stuff isn't (which is fine). Discovery and init are only supposed to be called from the main thread. Only the `Render()` method of the `MidiDeviceSoundCanvas` instances runs in a separate thread. The rest of `MidiDeviceSoundCanvas` and everything else CLAP related is in the same thread (main thread). But let me know if you see some potential threading issues; I can't see any.

Reply:
>  EnumeratePlugins() is private, so it's called only once. But there's no harm in calling it multiple times.

It's not safe to call `EnumeratePlugins()` once someone has a `Library` object because it calls the global init + deinit.

I didn't notice it was a private method though.  Your public method `GetPluginInfos()` ensures it only gets called once by caching the results.  So yeah, I was probably reading too much into this.  Your public methods look safe.

Reply:
> It's not safe to call `EnumeratePlugins()` once someone has a `Library` object because it calls the global init + deinit.

Ah, now I see it, cheers! `EnumeratePlugins()` should use `GetOrLoadLibrary()`, then it will be 100% correct and this cannot happen. Without this, cache invalidation could cause memory corruption with the current code (not with our plugin, but that is only resitant to this by fluke).

Making the change soon.

Reply:
Done.